pkg/resource/deploy/providers/registry.go

```diff
@@ -628,10 +628,24 @@ func (r *Registry) Diff(ctx context.Context, req plugin.DiffRequest) (plugin.Dif
 
 	// Diff the properties.
 	filteredNewInputs := FilterProviderConfig(req.NewInputs)
+	// Fixup the old inputs and outputs to contain "version", old versions of the CLI wouldn't have force saved it to
+	// state even though we know guarantee it's always in new inputs. This isn't _technically_ correct, we're just
+	// assuming the old state is the same version as the current state it might genuinely be different but there's no
+	// way for us to tell. But not restoring version can cause diffs even for the same version (see
+	// https://github.com/pulumi/pulumi-docker-build/issues/581).
+	fixup := func(pm resource.PropertyMap) resource.PropertyMap {
+		new := pm.Copy()
+		if _, has := new["version"]; !has {
+			new["version"] = req.NewInputs["version"]
+		}
+		return new
+	}
+	oldOutputs := fixup(req.OldOutputs)
+
 	diff, err := provider.DiffConfig(context.Background(), plugin.DiffConfigRequest{
 		URN:           req.URN,
-		OldInputs:     FilterProviderConfig(req.OldInputs),
-		OldOutputs:    req.OldOutputs, // OldOutputs is already filtered
+		OldInputs:     FilterProviderConfig(fixup(req.OldInputs)),
+		OldOutputs:    fixup(oldOutputs), // OldOutputs is already filtered
```

Comment:
Just to make sure I understand this correctly.  `OldInputs` and `OldOutputs` come from the old state, and thus may not have the version set.  `NewInputs` comes from the engine, and since https://github.com/pulumi/pulumi/pull/20313 always has the version set, and that's why things are wonky now?  And thus fixing up the version here if we don't have it to just be the same makes sense, because we only set it in the engine in `NewInputs`.

Reply:
ack